sdk/keyvault/azure-security-keyvault-jca/src/main/java/com/azure/security/keyvault/jca/implementation/utils/AccessTokenUtil.java

```diff
@@ -58,122 +75,249 @@ public final class AccessTokenUtil {
         LOGGER.entering("AccessTokenUtil", "getAccessTokenOnAppService", resource);
         LOGGER.info("Getting access token using managed identity based on MSI_SECRET");
+
         AccessToken result = null;
         StringBuilder url = new StringBuilder();
+
         url.append(System.getenv("MSI_ENDPOINT"))
            .append("?api-version=2017-09-01")
            .append(RESOURCE_FRAGMENT).append(resource);
+
         if (clientId != null) {
             url.append("&clientid=").append(clientId);
             LOGGER.log(INFO, "Using managed identity with client ID: {0}", clientId);
         }
 
         HashMap<String, String> headers = new HashMap<>();
+
         headers.put("Metadata", "true");
         headers.put("Secret", System.getenv("MSI_SECRET"));
+
         String body = HttpUtil.get(url.toString(), headers);
 
         if (body != null) {
             result = (AccessToken) JsonConverterUtil.fromJson(body, AccessToken.class);
         }
+
         LOGGER.exiting("AccessTokenUtil", "getAccessTokenOnAppService", result);
+
         return result;
     }
 
     /**
      * Get the authorization token on everything else but Azure App Service.
      *
-     * @param resource the resource.
-     * @param identity the user-assigned identity (null if system-assigned).
-     * @return the authorization token.
+     * @param resource The resource.
+     * @param identity The user-assigned identity (null if system-assigned).
+     * @return The authorization token.
      */
     private static AccessToken getAccessTokenOnOthers(String resource, String identity) {
         LOGGER.entering("AccessTokenUtil", "getAccessTokenOnOthers", resource);
         LOGGER.info("Getting access token using managed identity");
+
         if (identity != null) {
             LOGGER.log(INFO, "Using managed identity with object ID: {0}", identity);
         }
+
         AccessToken result = null;
 
         StringBuilder url = new StringBuilder();
+
         url.append(OAUTH2_MANAGED_IDENTITY_TOKEN_URL)
            .append(RESOURCE_FRAGMENT).append(resource);
+
         if (identity != null) {
             url.append("&object_id=").append(identity);
         }
 
         HashMap<String, String> headers = new HashMap<>();
+
         headers.put("Metadata", "true");
+
         String body = HttpUtil.get(url.toString(), headers);
 
         if (body != null) {
             result = (AccessToken) JsonConverterUtil.fromJson(body, AccessToken.class);
         }
+
         LOGGER.exiting("AccessTokenUtil", "getAccessTokenOnOthers", result);
+
         return result;
     }
+
+    public static String getLoginUri(URI resourceUri, boolean disableChallengeResourceVerification) {
+        HttpResponse response = HttpUtil
+            .postWithResponse(resourceUri.toString(), null, "", "application/x-www-form-urlencoded");
+        Map<String, String> challengeAttributes =
+            extractChallengeAttributes(response.getFirstHeader(WWW_AUTHENTICATE).getValue());
```

Comment:
This will throw a NPE if `HttpUtil.postWithResponse` returns `null` (which it does if the HTTP call fails).

Reply:
It seems like the pattern the library follows in other places like `HttpUtil.post()` or `HttpUtil.get()` that use `HttpUtil.createResponseHandler()`. Do you think we should change those as well?

Reply:
Maybe we can do that in a separate PR.

Reply:
I've changed this this to throw an `IllegalStateException` when no login URI can be retrieved.